syntax = "proto3";

package employees.v1;

option go_package = "github.com/opensourcecorp/workshops/protobuf-grpc/generated/go/employees/v1";
// TODO:
// option java_package = "org.opensourcecorp.workshops.protobuf_grpc.employees.v1";

// Allow for annotations/options to allow for HTTP transcoding of gRPC requests.
// You can spec & generate these protobufs just fine without, but to actually
// use them, you'll need to use the grpc-gateway plugin to take advantage of
// these annotations
import "google/api/annotations.proto";

// Service for interacting with an employee record database
service EmployeesService {
  // GetEmployee takes a short (friendly) name for a single employee, and
  // returns that employee's associated record
  rpc GetEmployee(GetEmployeeRequest) returns (GetEmployeeResponse) {
    // This is where you use those imported annotations above -- this maps GET
    // requests over HTTP to the path specified, for example
    option (google.api.http) = {
      get: "/employees/v1/get_employee/{short_name}"
    };
  }

  // ListEmployees returns all possible employees' short names that can be used
  // in calls to GetEmployee
  rpc ListEmployees(ListEmployeesRequest) returns (ListEmployeesResponse) {
    option (google.api.http) = {
      get: "/employees/v1/list_employees"
    };
  }
}

// For calling GetEmployee
message GetEmployeeRequest {
  string short_name = 1;
}

// Response from the GetEmployee call
message GetEmployeeResponse {
  // Contains Employee record details
  Employee employee = 1;
}

// Even for a Service that you don't have parameters specified for, it's good
// practice to have the parameters be an empty message like this, for
// forwards-compatibility in case you ever do add parameter details
message ListEmployeesRequest {}

// Response from the ListEmployees call
message ListEmployeesResponse {
  // List of short (friendly) names for Employees
  repeated string short_names = 1;
}

// Defines the record structure for each employee
message Employee {
  // Employee ID number
  int64 id = 1;
  // Employee full name
  string full_name = 2;
  // Employee birthday, in RFC 3339 date format
  string birthday = 3;
}
